src/lib/transition_frontier/persistent_frontier/database.ml

```diff
@@ -0,0 +1,330 @@
+open Async_kernel
+open Core
+open Coda_base
+open Coda_transition
+open Frontier_base
+
+let rec deferred_list_result_iter ls ~f =
+  let open Deferred.Result.Let_syntax in
+  match ls with
+  | [] ->
+      return ()
+  | h :: t ->
+      let%bind () = f h in
+      deferred_list_result_iter t ~f
+
+(* TODO: should debug assert garbage checks be added? *)
+open Result.Let_syntax
+
+(* TODO: implement versions with module versioning. For
+ * now, this is just stubbed so we can add db migrations
+ * later.
+ *)
+let version = 1
+
+module Schema = struct
+  module Keys = struct
+    module String = String
+
+    module Prefixed_state_hash = struct
+      type t = string * State_hash.Stable.V1.t [@@deriving bin_io]
+    end
+  end
+
+  type _ t =
+    | Db_version : int t
+    | Transition : State_hash.Stable.V1.t -> External_transition.Stable.V1.t t
+    | Arcs : State_hash.Stable.V1.t -> State_hash.Stable.V1.t list t
```

Comment:
Why do we now have the GADT type of `Arcs`? I think previously before, I merged `Arcs` and `Transition`s together.

Reply:
Because it's more efficient to update arcs this way. We don't have to deserialize and reserialize the whole external transition every time we want to append a state hash to it's arcs.